src/analysis/box_integral/box/basic.lean

```diff
@@ -0,0 +1,326 @@
+  tfae_have : 4 → 2, from λ h x hx i, Ioc_subset_Ioc (h.1 i) (h.2 i) (hx i),
+  tfae_finish
+end
+
+variables {I J}
+
+@[simp, norm_cast] lemma coe_subset_coe : (I : set (ι → ℝ)) ⊆ J ↔ I ≤ J := iff.rfl
+lemma le_iff_bounds : I ≤ J ↔ J.lower ≤ I.lower ∧ I.upper ≤ J.upper := (le_tfae I J).out 0 3
+
+lemma injective_coe : injective (coe : box ι → set (ι → ℝ)) :=
+begin
+  rintros ⟨l₁, u₁, h₁⟩ ⟨l₂, u₂, h₂⟩ h,
+  simp only [subset.antisymm_iff, coe_subset_coe, le_iff_bounds] at h,
+  congr,
+  exacts [le_antisymm h.2.1 h.1.1, le_antisymm h.1.2 h.2.2]
+end
+
+@[simp, norm_cast] lemma coe_inj : (I : set (ι → ℝ)) = J ↔ I = J :=
+injective_coe.eq_iff
+
+@[ext] lemma ext (H : ∀ x, x ∈ I ↔ x ∈ J) : I = J :=
+injective_coe $ set.ext H
+
+lemma ne_of_disjoint_coe (h : disjoint (I : set (ι → ℝ)) J) : I ≠ J :=
+mt coe_inj.2 $ h.ne I.coe_ne_empty
+
+instance : partial_order (box ι) :=
+{ le := (≤),
+  .. partial_order.lift (coe : box ι → set (ι → ℝ)) injective_coe }
+
+/-- Closed box corresponding to `I : box_integral.box ι`. -/
+protected def Icc : box ι ↪o set (ι → ℝ) :=
+order_embedding.of_map_le_iff (λ I : box ι, Icc I.lower I.upper) (λ I J, (le_tfae I J).out 2 0)
+
+lemma Icc_def : I.Icc = Icc I.lower I.upper := rfl
+
+@[simp] lemma upper_mem_Icc (I : box ι) : I.upper ∈ I.Icc := right_mem_Icc.2 I.lower_le_upper
+@[simp] lemma lower_mem_Icc (I : box ι) : I.lower ∈ I.Icc := left_mem_Icc.2 I.lower_le_upper
+
+protected lemma is_compact_Icc : is_compact I.Icc := is_compact_pi_Icc I.lower I.upper
+
+lemma Icc_eq_pi : I.Icc = pi univ (λ i, Icc (I.lower i) (I.upper i)) := (pi_univ_Icc _ _).symm
+
+lemma le_iff_Icc : I ≤ J ↔ I.Icc ⊆ J.Icc := (le_tfae I J).out 0 2
+
+lemma antitone_lower : antitone (λ I : box ι, I.lower) :=
+λ I J H, (le_iff_bounds.1 H).1
+
+lemma monotone_upper : monotone (λ I : box ι, I.upper) :=
+λ I J H, (le_iff_bounds.1 H).2
+
+lemma coe_subset_Icc : ↑I ⊆ I.Icc := λ x hx, ⟨λ i, (hx i).1.le, λ i, (hx i).2⟩
+
+lemma part_eq_part {o₁ o₂ : part (box ι)} :
+  o₁ = o₂ ↔ (⋃ J ∈ o₁, ↑J : set (ι → ℝ)) = (⋃ J ∈ o₂, ↑J : set (ι → ℝ)) :=
+begin
+  refine ⟨λ h, h ▸ rfl, λ h, _⟩,
+  rcases ⟨o₁.eq_none_or_eq_some, o₂.eq_none_or_eq_some⟩ with ⟨rfl|⟨I₁, rfl⟩, rfl|⟨I₂, rfl⟩⟩;
+    try { refl }; simpa using h
+end
+
+/-!
+### Intersection of two boxes
+
+Since two nonempty boxes can be disjoint, we can't define a `has_inf` instance on
+`box_integral.box ι`. So, we define a function `box_integral.box.inter` that takes a proof of
+`(I ∩ J : set (ι → ℝ)).nonempty` as an argument.
+-/
+
+lemma nonempty_coe_inter_coe {I J : box ι} :
+  (I ∩ J : set (ι → ℝ)).nonempty ↔ ∀ i, max (I.lower i) (J.lower i) < min (I.upper i) (J.upper i) :=
+by simp only [coe_eq_pi, ← pi_inter_distrib, univ_pi_nonempty_iff, Ioc_inter_Ioc, set.nonempty_Ioc,
+  sup_eq_max, inf_eq_min]
+
+/-- Intersection of two boxes. Since two nonempty boxes can be disjoint, this function that takes a
+proof of `(I ∩ J : set (ι → ℝ)).nonempty` as an argument. -/
```

Comment:
```suggestion
/-- Intersection of two boxes. Since two nonempty boxes can be disjoint, this function takes a
proof of `(I ∩ J : set (ι → ℝ)).nonempty` as an argument. -/
```